### ISSUE TITLE
Python pip installation

### ISSUE THREAD

**Issue by impredicative** [thomastrapp/hext#1]
It would really help if hext was pip-installable for Python. It would really simplify the installation and use. I mean to have it published as a package at [pypi](https://pypi.org/). This applies to both command-line and Python use.

This request is relevant because a lot of data analysis work is done via online notebooks, where only pip installation of packages is an option.

**Comment by thomastrapp** [comment]
Yes, hext badly needs a way to be installable without compiling from sources.
I will look into creating a python package.

**Comment by impredicative** [comment]
If it's simpler, a first step may be a routine apt-installable Ubuntu package. A pip package can come later. For now I have it working locally from source.

**Comment by brandonrobertz** [comment]
Another option is to simply release binaries with the github release mechanism. Lots of programs do this, including jq, which I think has the most relevance here: https://github.com/stedolan/jq/releases

**Comment by impredicative** [comment]
Ah. Releasing the binaries can be the first step, with the second step being the debian package. The benefit of the debian package is of course that it can auto-update using `apt`. pip installation is probably the most complicated since it requires Python 3 bindings too.

**Comment by thomastrapp** [comment]
Thank you both very much for your input, I really like this idea.

* libhext, statically linked with all (most) dependencies, especially Boost Regex and Gumbo
* htmlext, statically linked with Boost Program Options and libhext
* hext.so module for Python 3.6 (or an earlier version, if Python is backwards compatible), linked with the static version of libhext
* x86_64 only
* Clean Ubuntu 18.04 as a build host, because Ubuntu is both popular and able to compile the most recent version of hext out of the box. An alternative might be Debian Sid/"Buster".
* Add a file containing the checksums of each binary
* Add a signature for the checksum file that is verifiable with my public key. So it doesn't look too shady.

That might actually work. Fingers crossed for binary compatibility.

It's certainly not the cleanest approach to packaging, but a compromise until the Debian packages are ready. Additionally, this might be a way to support non-Debian based distributions.

Creating Debian packages shouldn't be too hard either. It's just that "doing it the right way" is a lot of work (especially as a noob) and is required for being accepted into the Debian archive. I will work on that as well, since getting a package into Debian has been on my wish list for a long time.

**Comment by impredicative** [comment]
A third-party ppa repository can perhaps be used until the official one is approved. This is very common and allows for updates.

**Comment by brandonrobertz** [comment]
@thomastrapp I tested the static binary and python binding and can confirm they work on my Ubuntu 18.04.1 LTS (x86_64).

**Comment by impredicative** [comment]
Although the deb package for hext worked for me, I'd still like to be able to install hext and its Python binding in a Python virtual environment from PyPI using pip. I need to use it with Python 3.7. Thanks.

**Comment by thomastrapp** [comment]
Thank you for your continued interest. I almost forgot the python package :)
I will put some hours into it this week.

**Comment by thomastrapp** [comment]
The Hext Python module is now installable through pip on x86_64:
```
pip install hext
```
Requires either Python 2.7, 3.4, 3.5, 3.6 or 3.7.

https://pypi.org/project/hext/0.1.0/#files
[hext-0.1.0-wheels.sha256sum.txt](https://github.com/thomastrapp/hext/files/2646419/hext-0.1.0-wheels.sha256sum.txt)
[hext-0.1.0-wheels.sha256sum.sig.txt](https://github.com/thomastrapp/hext/files/2646420/hext-0.1.0-wheels.sha256sum.sig.txt)

An easy way to test a pip install is to run the following:
```
pip install hext
git clone https://github.com/thomastrapp/hext.git
HTMLEXT="hext/libhext/bindings/python/htmlext.py" \
  ./hext/test/blackbox.sh \
  hext/test/case/*hext
```
So far, so good! :)

Todo:
* Add a PyPI package for the command-line utility
* Add a proper readme.md, with an introduction to using Hext with Python
* Update documentation

**Comment by thomastrapp** [comment]
The htmlext command-line utility is now installable through pip on x86_64:
```
pip install htmlext
```

https://pypi.org/project/htmlext/0.1.0/#files
[htmlext-0.1.0-wheels.sha256sum.txt](https://github.com/thomastrapp/hext/files/2648850/htmlext-0.1.0-wheels.sha256sum.txt)
[htmlext-0.1.0-wheels.sha256sum.asc.txt](https://github.com/thomastrapp/hext/files/2648851/htmlext-0.1.0-wheels.sha256sum.asc.txt)

**Comment by impredicative** [comment]
Does it have to be two different packages? Why can't both be bundled into one, i.e. into `hext`? Why make it complicated?

**Comment by thomastrapp** [comment]
Granted, I'm not familiar with how projects are usually packaged for pypi. But it made sense to me to provide two packages: The components cover different use-cases (CLI vs Python) and are independent of each other.

Also, the htmlext package installs a binary into the user's `$PATH` (usually). A user might want to use only the Python module, and might not expect or want this.

**Comment by impredicative** [comment]
OK; it's not an issue worth debating. Having said that, any pip-based installation will typically not have the `sudo` permissions needed to install anything outside the virtual environment that pip is being run under. For example, when I install pylint in my virtual environment, I can access the pylint CLI only from the virtual environment.

**Comment by brandonrobertz** [comment]
One example of a combined package providing a library and also an executable is [html2text](https://github.com/aaronsw/html2text). Virtualenv should automatically add the virtualenv's `bin/` directory to the path. For example, the html2text executable is on my path once I activate my virtualenv: `~/.virtualenvs/example/bin/html2text`.

**Comment by thomastrapp** [comment]
Looks like there's no reason for maintaining separate packages.
I'll move htmlext into hext.

**Comment by thomastrapp** [comment]
New release 0.2.0:
* `pip install hext` now installs htmlext as well
* Added a readme
* Bugfix: unicode strings can now be passed to hext.Hext() and hext.Html() when using python 2
* I deleted the htmlext package from pypi.org

[hext-0.2.0-wheels.sha256sum.txt](https://github.com/thomastrapp/hext/files/2655441/hext-0.2.0-wheels.sha256sum.txt)
[hext-0.2.0-wheels.sha256sum.asc.txt](https://github.com/thomastrapp/hext/files/2655442/hext-0.2.0-wheels.sha256sum.asc.txt)

If there are any issues, or if something looks out of place, please let me know.

@brandonrobertz 
> I've included the use of Hext in a research paper which is coming out next month.

I would love to see the paper when it's finished, if possible :)

**Comment by impredicative** [comment]
@thomastrapp I can now confirm that the `hext==0.2.0` PyPI package worked well for me with Python 3.7.1 on Ubuntu 18.04.1. Thanks so much!

**Comment by brandonrobertz** [comment]
Just a FYI @thomastrapp, I just had a PR merged in [Workbench](https://github.com/CJWorkbench/cjworkbench) to include the python Hext package by default. Everyone was real impressed with it. I'm working on a browser-based Hext template building tool ([found here](https://github.com/brandonrobertz/autoscrape-extractor-workbench)) which will be a simplified way for people to interactively build Hext templates. I can share the research paper privately (email).

**Comment by thomastrapp** [comment]
@brandonrobertz 
> I just had a PR merged in Workbench to include the python Hext package by default. Everyone was real impressed with it.

Very glad to hear that. Made my day :+1: 

> I'm working on a browser-based Hext template building tool (found here) which will be a simplified way for people to interactively build Hext templates.

Brilliant!
I used the current version to create a list of GCC releases:
https://app.workbenchdata.com/workflows/8582/

**Comment by impredicative** [comment]
@thomastrapp I am now using the Python hext==0.2.3 in my [project](https://github.com/impredicative/irc-rss-feed-bot) which serves hundreds of users. Its YAML config file includes `hext` strings as input for configurable scraping of web pages as virtual feeds. Thanks. Credit to brandonrobertz for the original recommendation.

**Comment by thomastrapp** [comment]
> I am now using the Python hext==0.2.3 in my project which serves hundreds of users

That's very nice to hear! Thank you for letting me know.
